File: Sources/DarkModeCore/UIViewController+DarkModeKit.m
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//  Licensed under the MIT License.
//

#import "UIViewController+DarkModeKit.h"
#import "UIView+DarkModeKit.h"

@import ObjectiveC;

@implementation UIViewController (DarkModeKit)

- (DMTraitCollection *)dmTraitCollection {
  if (@available(iOS 13, *)) {
    return [DMTraitCollection traitCollectionWithUITraitCollection:self.traitCollection];
  }
  return DMTraitCollection.overrideTraitCollection;
}

- (void)dmTraitCollectionDidChange:(DMTraitCollection *)previousTraitCollection {
  if (@available(iOS 13, *))
    return;

  [self setNeedsStatusBarAppearanceUpdate];
  [[self presentedViewController] dmTraitCollectionDidChange:previousTraitCollection];
  [[self childViewControllers] enumerateObjectsUsingBlock:^(__kindof UIViewController * _Nonnull child, NSUInteger idx, BOOL * _Nonnull stop) {
    [child dmTraitCollectionDidChange:previousTraitCollection];
  }];
  if ([self isViewLoaded])
    [[self view] dmTraitCollectionDidChange:previousTraitCollection];
}

@end

File: Sources/FluentDarkModeKit/Extensions/UIViewController+DarkModeKit.swift
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//  Licensed under the MIT License.
//

extension UIViewController: DMTraitEnvironment {
  public var dmTraitCollection: DMTraitCollection {
    if #available(iOS 13.0, *) {
      return DMTraitCollection(uiTraitCollection: traitCollection)
    }
    return DMTraitCollection.override
  }

  open func dmTraitCollectionDidChange(_ previousTraitCollection: DMTraitCollection?) {
    if #available(iOS 13.0, *) {
      return
    }

    setNeedsStatusBarAppearanceUpdate()
    presentedViewController?.dmTraitCollectionDidChange(previousTraitCollection)
    children.forEach { $0.dmTraitCollectionDidChange(previousTraitCollection) }
    if isViewLoaded {
      view.dmTraitCollectionDidChange(previousTraitCollection)
    }
  }
}
